# Problem 1

## 1a

$\newcommand{Vopt}[2]{V_\text{opt}^{(#1)}(#2)}$
We initialize $\Vopt{0}{s}=0$ for all $s$.
Then we apply value iteration:

$$
\Vopt{t+1}{s} = \max_{a\in\{-1,+1\}} \sum_{s'\in\{-2,\ldots,2\}} T(s,a,s')(\text{Reward}(s,a,s')+\Vopt{t}{s'})
$$

Here $\gamma=1$ and for simplicity I am not writing that $\Vopt{t}{s}=0$ for any end state $s$, i.e. for $s\in\{-2,2\}$.

Now to determine $\Vopt{1}{s}$.

- $\Vopt{1}{-1}$ is the expected utility of being in state $s=-1$ and applying the optimal policy.
    - The optimal policy would choose either $a=-1$ or $a=+1$ depending on which yields a greater expected utility. First consider playing $a=-1$.
        - We'd have a $T(-1,-1,-2)=0.8$ chance of earning $r=20$ and reaching state $s'=-2$, 
        - and we'd have a $T(-1,-1,0)=0.2$ chance of earning $r=-5$ and reaching state $s'=0$.
        - Thus the expected value of playing $a=-1$ 
          is 
          $$Q_\text{opt}^{(1)}(-1,-1) = 0.8 (20 + \Vopt{0}{-2}) + 0.2 (-5 + \Vopt{0}{0}) = 0.8(20)+0.2(-5) = 15$$
    - Now we consider playing $a=+1$.
        - We'd have a $T(-1,+1,-2)=0.7$ chance of earning $r=20$ and reaching state $s'=-2$, 
        - and we'd have a $T(-1,+1,0)=0.3$ chance of earning $r=-5$ and reaching state $s'=0$.
        - Thus the expected value of playing $a=+1$ 
          is 
          $$Q_\text{opt}^{(1)}(-1,+1) = 0.7 (20 + \Vopt{0}{-2}) + 0.3 (-5 + \Vopt{0}{0}) = 0.7(20)+0.3(-5) = 12.5$$
    - The optimal policy (according to what we think so far; remember we are biased by our initially crappy $V_\text{opt}$ estimate) would therefore play $a=-1$ when at state $s=-1$, which
      makes $\Vopt{1}{-1} = Q_\text{opt}^{(1)}(s=-1,a=-1) = 15$
- $\Vopt{1}{0}$
    - expected value of playing $a=-1$ is
      $$0.8 (-5 + \Vopt{0}{-1}) + 0.2 (-5 + \Vopt{0}{1}) = 0.8(-5)+0.2(-5) = -5$$
    - expected value of $a=+1$ is
      $$0.7 (-5 + \Vopt{0}{-1}) + 0.3 (-5 + \Vopt{0}{1}) = 0.7(-5)+0.3(-5) = -5$$
    - $a=-1$ and $a=+1$ are tied when $s=-1$. The max between $Q_\text{opt}^{(1)}(s=0,a=-1)$ and $Q_\text{opt}^{(1)}(s=0,a=+1)$ is $-5$. So
      $\Vopt{1}{0} = -5$.
- $\Vopt{1}{1}$
    - expected value of playing $a=-1$ is
      $$0.8 (-5 + \Vopt{0}{0}) + 0.2 (100 + \Vopt{0}{2}) = 0.8(-5)+0.2(100) = 16$$
    - expected value of $a=+1$ is
      $$0.7 (-5 + \Vopt{0}{0}) + 0.3 (100 + \Vopt{0}{2}) = 0.7(-5)+0.3(100) = 26.5$$
    - $a=+1$ is optimal
      $\Vopt{1}{1} = Q_\text{opt}^{(1)}(s=1,a=+1) = 26.5$.     

We ended up with $\Vopt{1}{s}$ being $0,15,-5,26.5,0$ for $s=-2,1,0,1,2$ respectively.

Another iteration, let's determine $\Vopt{2}{s}$.

- $\Vopt{2}{-1}$
    - expected value of playing $a=-1$ is
      $$0.8 (20 + \Vopt{1}{-2}) + 0.2 (-5 + \Vopt{1}{0}) = 0.8(20)+0.2(-5-5) = 14$$
    - expected value of $a=+1$ is
      $$0.7 (20 + \Vopt{1}{-2}) + 0.3 (-5 + \Vopt{1}{0}) = 0.7(20)+0.3(-5-5) = 11$$
    - $\Vopt{2}{-1} = 14$.
- $\Vopt{2}{0}$
    - expected value of playing $a=-1$ is
      $$0.8 (-5 + \Vopt{1}{-1}) + 0.2 (-5 + \Vopt{1}{1}) = 0.8(-5+15)+0.2(-5+26.5) = 12.3$$
    - expected value of $a=+1$ is
      $$0.7 (-5 + \Vopt{1}{-1}) + 0.3 (-5 + \Vopt{1}{1}) = 0.7(-5+15)+0.3(-5+26.5) = 13.45$$
    - $\Vopt{2}{0} = 13.45$.
- $\Vopt{2}{1}$
    - expected value of playing $a=-1$ is
      $$0.8 (-5 + \Vopt{1}{0}) + 0.2 (100 + \Vopt{1}{2}) = 0.8(-5-5)+0.2(100) = 12$$
    - expected value of $a=+1$ is
      $$0.7 (-5 + \Vopt{1}{0}) + 0.3 (100 + \Vopt{1}{2}) = 0.7(-5-5)+0.3(100) = 23$$
    - $\Vopt{2}{1} = 23$.

We ended up with $\Vopt{2}{s}$ being $0,14,13.45,23,0$ for $s=-2,1,0,1,2$ respectively.

## 1b

Let's do one more iteration, but this time just focusing on the $Q$s so that we can determine the optimal policy.

- From $s=-1$
  - $Q_\text{opt}^{(3)}(s=-1,a=-1) = 0.8 (20 + \Vopt{2}{-2}) + 0.2 (-5 + \Vopt{2}{0}) $ 
  `= 0.8*(20)+0.2*(-5+13.45) = 17.69`
  - $Q_\text{opt}^{(3)}(s=-1,a=1) = 0.7 (20 + \Vopt{2}{-2}) + 0.3 (-5 + \Vopt{2}{0}) $ 
  `= 0.7*(20)+0.3*(-5+13.45) = 16.535`
- From $s=0$
  - $Q_\text{opt}^{(3)}(s=0,a=-1) = 0.8 (-5 + \Vopt{2}{-1}) + 0.2 (-5 + \Vopt{2}{1}) $ 
  `= 0.8*(-5 + 14)+0.2*(-5+23) = 10.8`
  - $Q_\text{opt}^{(3)}(s=0,a=1) = 0.7 (-5 + \Vopt{2}{-1}) + 0.3 (-5 + \Vopt{2}{1}) $ 
  `= 0.7*(-5+14)+0.3*(-5+23) = 11.7`
- From $s=1$
  - $Q_\text{opt}^{(3)}(s=1,a=-1) = 0.8 (-5 + \Vopt{2}{0}) + 0.2 (100 + \Vopt{2}{2}) $ 
  `= 0.8*(-5+13.45)+0.2*(100) = 26.76`
  - $Q_\text{opt}^{(3)}(s=1,a=1) = 0.7 (-5 + \Vopt{2}{0}) + 0.3 (100 + \Vopt{2}{2}) $ 
  `= 0.7*(-5+13.45)+0.3*(100) = 35.915`

So here's our optimal policy based on the estimate $\Vopt{2}{s}$ of $V_\text{opt}(s)$:
- $\pi(-1) = -1$
- $\pi(0)  = +1$
- $\pi(1)  = +1$

# Problem 2

## 2a
Testing that my solution in submission.py works:

In [1]:
from util import *
from submission import *

vi = ValueIteration()
vi.solve(CounterexampleMDP())
vi.V

In [2]:
# This class is not written by me, it is copied from grader.py
class AddNoiseMDP(util.MDP):
    def __init__(self, originalMDP):
        self.originalMDP = originalMDP

    def startState(self):
        return self.originalMDP.startState()

    # Return set of actions possible from |state|.
    def actions(self, state):
        return self.originalMDP.actions(state)

    # Return a list of (newState, prob, reward) tuples corresponding to edges
    # coming out of |state|.
    def succAndProbReward(self, state, action):
        originalSuccAndProbReward = self.originalMDP.succAndProbReward(state, action)
        newSuccAndProbReward = []
        for state, prob, reward in originalSuccAndProbReward:
            newProb = 0.5 * prob + 0.5 / len(originalSuccAndProbReward)
            newSuccAndProbReward.append((state, newProb, reward))
        return newSuccAndProbReward

    # Return set of actions possible from |state|.
    def discount(self):
        return self.originalMDP.discount()

    
    
vi = ValueIteration()
vi.solve(AddNoiseMDP(CounterexampleMDP()))
vi.V


ValueIteration: 2 iterations


{'happy': 0, 'sad': 0, 'begin': -49.0}

(indeed the value of "begin" state has increased)

# 2b
If the MDP graph is acyclic then we should be able to apply the recurrence directly in an algorithm:

$$
\begin{align*}
&V_\text{opt}(s) = 0\hspace{2em} \text{if $s$ is an end state, otherwise:}\\
&V_\text{opt}(s) = \max_{a\in\text{Actions}(s)} \sum_{s'\in \text{States}\\T(s,a,s')\neq 0} T(s,a,s')(\text{Reward}(s,a,s')+\gamma V_\text{opt}(s'))
\end{align*}
$$

The acyclicity condition guarantees that this terminates; note that we only consider terms in the sum for which $T(s,a,s')$ is nonzero.

We can use memoization: cache the already-computed $V_\text{opt}(s)$ values. I think this guarantees that we won't pass over any $(s,a,s')$ triple more than once.

## 2c

For short I will write $R$ for $\text{Reward}$ and $V$ for $V_\text{opt}$. I'll use $V'$ for $V'_\text{opt}$, the optimal policy value function for the new MDP. The goal is to define $T'$ and $R'$ so that $V'=V$ on states where both are defined (i.e. states other than $o$).



We will make the new state $o$ an end state, so that $V'(o)=0$. There will never be a reward for going to the state $o$, so $R'(s,a,o)=0$ for all $s\in\text{States}$ and $a\in\text{Actions}(s)$.

For $s,s'\in\text{States}$ and $a\in\text{Actions}(s)$, we define

$$\begin{align*}
T'(s,a,s') &= \gamma \,T(s,a,s')\\
R'(s,a,s') &= \tfrac{1}{\gamma} R(s,a,s')
\end{align*}$$

We've now arranged things so that for $s,s'\in\text{States}$ and $a\in\text{Actions}(s)$,

$$\begin{align*}
T'(s,a,s')(R'(s,a,s') + v) = T(s,a,s')(R(s,a,s') + \gamma v)
\end{align*}$$

for any real number $v$.

We still have to define $T'(s,a,o)$, but this is forced on us by the requirement that the $T$'s be probabilities:

$$\begin{align*}
T'(s,a,o) = 1-\gamma
\end{align*}$$

for all $s\in\text{States}$ and $a\in\text{Actions}(s)$.

Now the recursion relation that $V'$ satisfies is

$$\begin{align*}
V'(s) &= 0\hspace{2em} \text{if $s$ is an end state, otherwise:}\\
V'(s) &= \max_{a\in\text{Actions}(s)} \sum_{s'\in \text{States'}} T'(s,a,s')(R'(s,a,s')+V'(s'))\\
&= \max_{a\in\text{Actions}(s)} \left[\left(\sum_{s'\in \text{States}} T(s,a,s')(R(s,a,s')+\gamma V'(s'))\right) + T'(s,a,o)(0+0)\right]\\
&= \max_{a\in\text{Actions}(s)} \sum_{s'\in \text{States}} T(s,a,s')(R(s,a,s')+\gamma V'(s'))
\end{align*}$$

This is exactly the same recursion relation that $V(s)$ satisfies, and so $V'(s)=V(s)$.

---

It's interesting that discounting the future by factor $\gamma$ is equivalent to having a probability $1-\gamma$ of terminating at each action. If it weren't for death, humans probably would value future rewards more.

# Problem 3a

Testing out the `BlackjackMDP` I came up with:

In [3]:
from submission import *

In [4]:
mdp1 = BlackjackMDP(cardValues=[1, 5], multiplicity=2,
                                   threshold=10, peekCost=1)
preBustState = (6, None, (1, 1))

mdp1.succAndProbReward(preBustState,'Take')

[((7, None, (0, 1)), 0.5, 0), ((11, None, None), 0.5, 0)]

# 3b

In [93]:
vi = ValueIteration()
vi.solve(BlackjackMDP(
        cardValues = [3,4,18],
        multiplicity = 3,
        threshold = 20,
        peekCost = 1
        ))
#for key,val in vi.pi.items():
#    print("{: >20}{: >7}".format(str(key),str(val)))
print("Peek is chosen for",
      100*float(len([_ for _,val in vi.pi.items() if val=='Peek']))/len(vi.pi),
      "percent of the states.")

ValueIteration: 11 iterations
Peek is chosen for 13.793103448275861 percent of the states.


# 4b

In [1]:
import submission
import util
import numpy as np



def simulate_QL_over_MDP(mdp, featureExtractor,explorationProb,verbose=True):

    qlearner=submission.QLearningAlgorithm(
      actions=lambda s : ['Take','Quit','Peek'] if s[1] is None else ['Take','Quit'],
      discount=mdp.discount(), # Not really cheating I think
      featureExtractor=featureExtractor,
      explorationProb=explorationProb,
    )
    
    # Learn a policy
    rwds = util.simulate(
                          mdp=mdp,
                          rl=qlearner,
                          numTrials=30000,
                          maxIterations=10000,
                          verbose=False
                        )

    qlearner_ave_rwd_exploration = np.array(rwds).mean()

    # No more exploring, time to perform
    qlearner.explorationProb = 0.0

    # Perform
    rwds = util.simulate(
                          mdp=mdp,
                          rl=qlearner,
                          numTrials=30000,
                          maxIterations=10000,
                          verbose=False
                        )

    qlearner_ave_rwd = np.array(rwds).mean()

    vi = util.ValueIteration()
    vi.solve(mdp)
    vopt_start = vi.V[mdp.startState()]

    print(
        "Q Learner's average reward during exploration:",
        qlearner_ave_rwd_exploration,
        "\nQ Learner's average reward after exploration:",
        qlearner_ave_rwd,
        "\n Expected utility of true optimal policy:",
        vopt_start
         )
    
    states_with_diff_actions = [s for s in mdp.states if vi.pi[s]!=qlearner.getAction(s)]
    num_early_quits = len([s for s in mdp.states if vi.pi[s]!=qlearner.getAction(s) and qlearner.getAction(s) == "Quit"])
    num_bad_takes   = len([s for s in mdp.states if vi.pi[s]!=qlearner.getAction(s) and qlearner.getAction(s) == "Take"])
    
    print(
        len(states_with_diff_actions),
        "out of",
        len(mdp.states),
        "states led to non-optimal actions.\n",
        num_early_quits,
        "of these are situations where the qlearner Quits before it should.\n",
        num_bad_takes,
        "of them are situations where the qlearner Takes when it shouldn't."
    )
    
    if verbose:
        print("Here are all the nonoptimal actions:")
        for s in states_with_diff_actions:
            print(
                "\tIn state",
                s,
                "our qlearner chooses to",
                qlearner.getAction(s),
                "when the optimal choice is to",
                vi.pi[s]
            )

In [11]:
simulate_QL_over_MDP(submission.smallMDP, submission.identityFeatureExtractor,0.3)

ValueIteration: 5 iterations
Q Learner's average reward during exploration: 4.6805 
Q Learner's average reward after exploration: 6.827566666666667 
 Expected utility of true optimal policy: 6.833333333333333
2 out of 27 states led to non-optimal actions.
 2 of these are situations where the qlearner Quits before it should.
 0 of them are situations where the qlearner Takes when it shouldn't.
Here are all the nonoptimal actions:
	In state (5, 1, (2, 1)) our qlearner chooses to Quit when the optimal choice is to Take
	In state (5, 0, (2, 1)) our qlearner chooses to Quit when the optimal choice is to Take


In [12]:
simulate_QL_over_MDP(submission.largeMDP, submission.identityFeatureExtractor,0.3,verbose = False)

ValueIteration: 15 iterations
Q Learner's average reward during exploration: 6.5995333333333335 
Q Learner's average reward after exploration: 8.769266666666667 
 Expected utility of true optimal policy: 35.52569097569097
946 out of 2745 states led to non-optimal actions.
 228 of these are situations where the qlearner Quits before it should.
 718 of them are situations where the qlearner Takes when it shouldn't.


What went wrong in the case of the large MDP? My first thought is that the model complexity is too high and there are just too many (state,action) pairs that are never visited by the Q Learner.

But it's not really that crazy... there are 2745 states and 3 actions, for a total of 8235 (state,action) pairs.
Why would 200K episodes (which I tried) not be sufficient to explore them?
Even the totally random exploration seems to not work:

In [7]:
simulate_QL_over_MDP(submission.largeMDP, submission.identityFeatureExtractor,1.0,verbose = False)

ValueIteration: 15 iterations
Q Learner's average reward during exploration: 4.486 
Q Learner's average reward after exploration: 11.3165 
 Expected utility of true optimal policy: 35.52569097569097
1739 out of 2745 states led to non-optimal actions.
 1334 of these are situations where the qlearner Quits before it should.
 405 of them are situations where the qlearner Takes when it shouldn't.


## 4c
Now we try a lower dimensional hand-crafted feature extractor:

In [2]:
simulate_QL_over_MDP(submission.smallMDP, submission.blackjackFeatureExtractor,0.3)

ValueIteration: 5 iterations
Q Learner's average reward during exploration: 4.839533333333334 
Q Learner's average reward after exploration: 6.828966666666667 
 Expected utility of true optimal policy: 6.833333333333333
5 out of 27 states led to non-optimal actions.
 3 of these are situations where the qlearner Quits before it should.
 0 of them are situations where the qlearner Takes when it shouldn't.
Here are all the nonoptimal actions:
	In state (10, None, None) our qlearner chooses to Quit when the optimal choice is to Take
	In state (7, None, None) our qlearner chooses to Peek when the optimal choice is to Take
	In state (6, 0, (1, 1)) our qlearner chooses to Quit when the optimal choice is to Take
	In state (1, None, None) our qlearner chooses to Peek when the optimal choice is to Take
	In state (6, None, None) our qlearner chooses to Quit when the optimal choice is to Take


In [3]:
simulate_QL_over_MDP(submission.largeMDP, submission.blackjackFeatureExtractor,0.3,verbose=False)

ValueIteration: 15 iterations
Q Learner's average reward during exploration: 19.1484 
Q Learner's average reward after exploration: 33.932966666666665 
 Expected utility of true optimal policy: 35.52569097569097
539 out of 2745 states led to non-optimal actions.
 521 of these are situations where the qlearner Quits before it should.
 10 of them are situations where the qlearner Takes when it shouldn't.


## 4d

In [12]:
import submission
import util
import numpy as np

vi = util.ValueIteration()
vi.solve(submission.originalMDP)
vi2 = util.ValueIteration()
vi2.solve(submission.newThresholdMDP)
rwds = util.simulate(
    mdp = submission.newThresholdMDP,
    rl = util.FixedRLAlgorithm(vi.pi),
    numTrials=30000
)
print("Expected reward from old policy: {}".format(np.array(rwds).mean()))
print("Expected reward from true optimal policy: {}".format(vi2.V[submission.newThresholdMDP.startState()]))
print("Now let's simulate with the q learner on the new policy...")
rwds = util.simulate(
    mdp = submission.newThresholdMDP,
    rl = submission.QLearningAlgorithm(
      actions=lambda s : ['Take','Quit','Peek'] if s[1] is None else ['Take','Quit'],
      discount=submission.newThresholdMDP.discount(), # Not really cheating I think
      featureExtractor=submission.blackjackFeatureExtractor,
    ),
    numTrials=30000
)
print("Expected reward from Q-learning behavior on new polocy: {}".format(np.array(rwds).mean()))

ValueIteration: 5 iterations
ValueIteration: 5 iterations
Expected reward from old policy: 6.8395
Expected reward from true optimal policy: 12.0
Now let's simulate with the q learner on the new policy...
Expected reward from Q-learning behavior on new polocy: 9.360066666666667
